File: ts/packages/sdk/src/index-config.ts
import type {
  ArtifactIndexSource,
  EmbedderConfig,
  EmbeddingsIndexConfig,
  EnrichmentConfig,
  GraphIndexSource,
  IndexConfig,
} from "./types.js";

const MAX_ARTIFACT_SOURCES = 64;

function cloneJsonValue(value: unknown, path: string): unknown {
  if (value === null || typeof value === "string" || typeof value === "boolean") return value;
  if (typeof value === "number") {
    if (!Number.isFinite(value)) throw new TypeError(`${path} must be finite`);
    return value;
  }
  if (Array.isArray(value)) {
    return value.map((item, index) => cloneJsonValue(item, `${path}[${index}]`));
  }
  if (typeof value === "object") {
    const prototype = Object.getPrototypeOf(value);
    if (prototype !== Object.prototype && prototype !== null) {
      throw new TypeError(`${path} must contain only JSON values`);
    }
    return Object.fromEntries(
      Object.entries(value).map(([key, item]) => [key, cloneJsonValue(item, `${path}.${key}`)])
    );
  }
  throw new TypeError(`${path} must contain only JSON values`);
}

function validateArtifactNames(artifacts: readonly string[]): void {
  if (artifacts.length === 0) throw new TypeError("at least one artifact source is required");
  if (artifacts.length > MAX_ARTIFACT_SOURCES) {
    throw new RangeError(`at most ${MAX_ARTIFACT_SOURCES} artifact sources are allowed`);
  }
  const seen = new Set<string>();
  artifacts.forEach((artifact, index) => {
    if (artifact.length === 0) throw new TypeError(`artifacts[${index}] is required`);
    if (seen.has(artifact))
      throw new TypeError(`duplicate artifact source ${JSON.stringify(artifact)}`);
    seen.add(artifact);
  });
}

/** Builds the shared artifact-only source shape used by full-text and vector indexes. */
export function artifactIndexSources(...artifacts: string[]): ArtifactIndexSource[] {
  validateArtifactNames(artifacts);
  return artifacts.map((artifact) => ({ artifact }));
}

/** Builds a full-text index over generated chunk or textual asset streams. */
export function artifactFullTextIndexConfig(name: string, ...artifacts: string[]): IndexConfig {
  if (name.length === 0) throw new TypeError("index name is required");
  return { name, type: "full_text", sources: artifactIndexSources(...artifacts) };
}

/**
 * Validates and copies ordered graph sources. Earlier sources win when more
 * than one source materializes the same edge identity.
 */
export function graphIndexSources(...sources: GraphIndexSource[]): GraphIndexSource[] {
  validateArtifactNames(sources.map((source) => source.artifact));
  sources.forEach((source, index) => {
    if (
      source.format !== undefined &&
      source.format !== "extraction_relation" &&
      source.format !== "extraction_graph"
    ) {
      throw new TypeError(`sources[${index}].format is invalid`);
    }
    if (
      source.nodes?.model !== undefined &&
      source.nodes.model !== "document" &&
      source.nodes.model !== "external"
    ) {
      throw new TypeError(`sources[${index}].nodes.model is invalid`);
    }
    for (const [fieldName, value] of [
      ["type", source.edge?.type],
      ["weight", source.edge?.weight],
    ] as const) {
      if (typeof value === "number" && !Number.isFinite(value)) {
        throw new TypeError(`sources[${index}].edge.${fieldName} must be finite`);
      }
    }
    const docFields = source.context?.doc_fields;
    if (docFields !== undefined) {
      if (docFields.some((field) => typeof field !== "string" || field.length === 0)) {
        throw new TypeError(
          `sources[${index}].context.doc_fields entries must be non-empty strings`
        );
      }
      if (new Set(docFields).size !== docFields.length) {
        throw new TypeError(`sources[${index}].context.doc_fields must be unique`);
      }
    }
  });
  return sources.map((source) => ({
    ...source,
    nodes: source.nodes === undefined ? undefined : { ...source.nodes },
    edge:
      source.edge === undefined
        ? undefined
        : {
            ...source.edge,
            metadata:
              source.edge.metadata === undefined
                ? undefined
                : (cloneJsonValue(
                    source.edge.metadata,
                    "graph source edge metadata"
                  ) as NonNullable<GraphIndexSource["edge"]>["metadata"]),
          },
    context:
      source.context === undefined
        ? undefined
        : {
            ...source.context,
            doc_fields:
              source.context.doc_fields === undefined ? undefined : [...source.context.doc_fields],
          },
  }));
}

export interface ArtifactEmbeddingSourceConfig {
  /** Stable name of the generated embedding artifact. */
  artifact: string;
  /** Optional upstream chunk artifact consumed by this embedding enrichment. */
  sourceArtifact?: string;
  /** Source text field. Defaults to `text`. */
  field?: string;
  /** Optional producer template; mutually exclusive with an empty field. */
  template?: string;
}

interface ArtifactEmbeddingIndexOptionsBase {
  sources: ArtifactEmbeddingSourceConfig[];
  embedder: EmbedderConfig;
}

export type ArtifactEmbeddingIndexOptions = ArtifactEmbeddingIndexOptionsBase &
  (
    | {
        /** Dense is the default. Omit dimension when the server can probe it. */
        sparse?: false;
        dimension?: number;
        distanceMetric?: NonNullable<EmbeddingsIndexConfig["distance_metric"]>;
      }
    | {
        /** Sparse indexes reject dense-only dimension and distance settings at compile time. */
        sparse: true;
        dimension?: never;
        distanceMetric?: never;
      }
  );

interface RuntimeArtifactEmbeddingIndexOptions extends ArtifactEmbeddingIndexOptionsBase {
  dimension?: number;
  sparse?: boolean;
  distanceMetric?: NonNullable<EmbeddingsIndexConfig["distance_metric"]>;
}

/**
 * Builds a multi-source vector index and its table-level embedding
 * enrichments through one typed SDK path.
 */
export function artifactEmbeddingIndexConfig(
  name: string,
  options: ArtifactEmbeddingIndexOptions
): IndexConfig {
  // Keep runtime checks for JavaScript consumers and values crossing an
  // untyped boundary even though TypeScript callers get a discriminated union.
  const runtimeOptions: RuntimeArtifactEmbeddingIndexOptions = options;
  if (name.length === 0) throw new TypeError("index name is required");
  validateArtifactNames(runtimeOptions.sources.map((source) => source.artifact));
  if (runtimeOptions.sparse && runtimeOptions.dimension !== undefined) {
    throw new TypeError("dimension must be omitted for sparse embedding indexes");
  }
  if (runtimeOptions.sparse && runtimeOptions.distanceMetric !== undefined) {
    throw new TypeError("distanceMetric must be omitted for sparse embedding indexes");
  }
  if (
    runtimeOptions.dimension !== undefined &&
    (!Number.isInteger(runtimeOptions.dimension) || runtimeOptions.dimension <= 0)
  ) {
    throw new RangeError("dimension must be a positive integer");
  }
  if (
    typeof (runtimeOptions.embedder as { provider?: unknown }).provider !== "string" ||
    (runtimeOptions.embedder as { provider: string }).provider.length === 0
  ) {
    throw new TypeError("embedder.provider is required");
  }
  if (
    runtimeOptions.distanceMetric !== undefined &&
    runtimeOptions.distanceMetric !== "l2_squared" &&
    runtimeOptions.distanceMetric !== "inner_product" &&
    runtimeOptions.distanceMetric !== "cosine"
  ) {
    throw new TypeError("distanceMetric is invalid");
  }

  const sources = artifactIndexSources(...runtimeOptions.sources.map((source) => source.artifact));
  const enrichments: EnrichmentConfig[] = runtimeOptions.sources.map((source, index) => {
    const field = source.field ?? "text";
    if (field.length === 0 && (source.template?.length ?? 0) === 0) {
      throw new TypeError(`sources[${index}] requires field or template`);
    }
    if (source.sourceArtifact === "") {
      throw new TypeError(`sources[${index}].sourceArtifact cannot be empty`);
    }
    return {
      name: source.artifact,
      kind: "embedding",
      ...(field.length > 0 ? { field } : {}),
      ...(source.template !== undefined ? { template: source.template } : {}),
      ...(source.sourceArtifact !== undefined
        ? { source_artifact_name: source.sourceArtifact }
        : {}),
      ...(runtimeOptions.dimension !== undefined ? { expected_dims: runtimeOptions.dimension } : {}),
    };
  });

  return {
    name,
    type: "embeddings",
    sources,
    enrichments,
    embedder: runtimeOptions.embedder,
    ...(runtimeOptions.sparse ? { sparse: true } : {}),
    ...(runtimeOptions.dimension !== undefined ? { dimension: runtimeOptions.dimension } : {}),
    ...(runtimeOptions.distanceMetric !== undefined
      ? { distance_metric: runtimeOptions.distanceMetric }
      : {}),
  };
}

File: py/packages/sdk/src/antfly/index_config.py
"""Validated helpers for artifact-backed index configuration."""

from __future__ import annotations

from collections.abc import Mapping, Sequence
from dataclasses import dataclass
from math import isfinite
from typing import Any, Literal

MAX_ARTIFACT_SOURCES = 64


def _validate_artifacts(artifacts: Sequence[object]) -> None:
    if not artifacts:
        raise ValueError("at least one artifact source is required")
    if len(artifacts) > MAX_ARTIFACT_SOURCES:
        raise ValueError(f"at most {MAX_ARTIFACT_SOURCES} artifact sources are allowed")
    seen: set[str] = set()
    for index, artifact in enumerate(artifacts):
        if not isinstance(artifact, str) or not artifact:
            raise ValueError(f"artifacts[{index}] is required")
        if artifact in seen:
            raise ValueError(f"duplicate artifact source {artifact!r}")
        seen.add(artifact)


def artifact_index_sources(*artifacts: str) -> list[dict[str, str]]:
    """Build the shared artifact-only source shape for full-text/vector indexes."""

    _validate_artifacts(artifacts)
    return [{"artifact": artifact} for artifact in artifacts]


def artifact_full_text_index_config(name: str, *artifacts: str, mem_only: bool = False) -> dict[str, Any]:
    """Build a full-text index over generated chunk or textual asset streams."""

    if not name:
        raise ValueError("index name is required")
    result: dict[str, Any] = {
        "name": name,
        "type": "full_text",
        "sources": artifact_index_sources(*artifacts),
    }
    if mem_only:
        result["mem_only"] = True
    return result


GraphArtifactFormat = Literal["extraction_relation", "extraction_graph"]
GraphNodeModel = Literal["document", "external"]


def _clone_json_value(value: Any, path: str) -> Any:
    if value is None or isinstance(value, (str, bool)):
        return value
    if isinstance(value, (int, float)) and not isinstance(value, bool):
        if not isfinite(value):
            raise ValueError(f"{path} must be finite")
        return value
    if isinstance(value, Mapping):
        result: dict[str, Any] = {}
        for key, child in value.items():
            if not isinstance(key, str):
                raise ValueError(f"{path} keys must be strings")
            result[key] = _clone_json_value(child, f"{path}.{key}")
        return result
    if isinstance(value, Sequence) and not isinstance(value, (str, bytes, bytearray)):
        return [_clone_json_value(child, f"{path}[{index}]") for index, child in enumerate(value)]
    raise ValueError(f"{path} must contain only JSON values")


@dataclass(frozen=True, slots=True)
class GraphNodeMapping:
    """Optional node identifier templates for one graph artifact stream."""

    model: GraphNodeModel = "document"
    source: str | None = None
    target: str | None = None


@dataclass(frozen=True, slots=True)
class GraphEdgeMapping:
    """Optional edge templates and metadata for one graph artifact stream."""

    type: str | int | float | None = None
    weight: str | int | float | None = None
    metadata: Mapping[str, Any] | None = None


@dataclass(frozen=True, slots=True)
class GraphContextMapping:
    """Document fields explicitly exposed to graph mapping templates."""

    doc_fields: Sequence[str]


@dataclass(frozen=True, slots=True)
class GraphArtifactSource:
    """One graph artifact stream and its source-specific payload interpretation."""

    artifact: str
    path: str | None = None
    format: GraphArtifactFormat = "extraction_relation"
    mention_edge_type: str | None = None
    nodes: GraphNodeMapping | None = None
    edge: GraphEdgeMapping | None = None
    context: GraphContextMapping | None = None


def graph_index_sources(*sources: GraphArtifactSource) -> list[dict[str, Any]]:
    """Build validated graph sources without accepting unknown source fields."""

    _validate_artifacts([source.artifact for source in sources])
    result: list[dict[str, Any]] = []
    for index, source in enumerate(sources):
        if source.format not in ("extraction_relation", "extraction_graph"):
            raise ValueError(f"sources[{index}].format is invalid")
        if source.nodes is not None and source.nodes.model not in ("document", "external"):
            raise ValueError(f"sources[{index}].nodes.model is invalid")
        if source.edge is not None:
            for field_name, value in (("type", source.edge.type), ("weight", source.edge.weight)):
                if value is not None and (isinstance(value, bool) or not isinstance(value, (str, int, float))):
                    raise ValueError(f"sources[{index}].edge.{field_name} must be a string or number")
                if isinstance(value, (int, float)) and not isinstance(value, bool) and not isfinite(value):
                    raise ValueError(f"sources[{index}].edge.{field_name} must be finite")
        if source.context is not None:
            fields = list(source.context.doc_fields)
            if any(not isinstance(field, str) or not field for field in fields):
                raise ValueError(f"sources[{index}].context.doc_fields entries must be non-empty strings")
            if len(fields) != len(set(fields)):
                raise ValueError(f"sources[{index}].context.doc_fields must be unique")
        item: dict[str, Any] = {
            "artifact": source.artifact,
            "format": source.format,
        }
        if source.path is not None:
            item["path"] = source.path
        if source.mention_edge_type is not None:
            item["mention_edge_type"] = source.mention_edge_type
        if source.nodes is not None:
            item["nodes"] = {
                key: value
                for key, value in {
                    "model": source.nodes.model,
                    "source": source.nodes.source,
                    "target": source.nodes.target,
                }.items()
                if value is not None
            }
        if source.edge is not None:
            item["edge"] = {
                key: value
                for key, value in {
                    "type": source.edge.type,
                    "weight": source.edge.weight,
                    "metadata": (
                        _clone_json_value(source.edge.metadata, f"sources[{index}].edge.metadata")
                        if source.edge.metadata is not None
                        else None
                    ),
                }.items()
                if value is not None
            }
        if source.context is not None:
            item["context"] = {"doc_fields": list(source.context.doc_fields)}
        result.append(item)
    return result


@dataclass(frozen=True, slots=True)
class ArtifactEmbeddingSource:
    """One embedding enrichment and the artifact name consumed by an index."""

    artifact: str
    source_artifact: str | None = None
    field: str = "text"
    template: str | None = None


def _model_dict(value: Mapping[str, Any] | Any) -> dict[str, Any]:
    if isinstance(value, Mapping):
        return dict(value)
    to_dict = getattr(value, "to_dict", None)
    if callable(to_dict):
        result = to_dict()
        if isinstance(result, dict):
            return result
    raise TypeError("embedder must be a mapping or generated SDK model")


def artifact_embedding_index_config(
    name: str,
    *,
    sources: Sequence[ArtifactEmbeddingSource],
    embedder: Mapping[str, Any] | Any,
    dimension: int | None = None,
    sparse: bool = False,
    distance_metric: str | None = None,
    vector_space: str | None = None,
) -> dict[str, Any]:
    """Build one vector-index config plus its embedding enrichments.

    Omitting ``vector_space`` selects Antfly's automatic semantic-producer
    validation. Set it only to assert compatibility across distinct producers.
    """

    if not name:
        raise ValueError("index name is required")
    _validate_artifacts([source.artifact for source in sources])
    if sparse and dimension is not None:
        raise ValueError("dimension must be omitted for sparse embedding indexes")
    if sparse and distance_metric is not None:
        raise ValueError("distance_metric must be omitted for sparse embedding indexes")
    if dimension is not None and (not isinstance(dimension, int) or isinstance(dimension, bool) or dimension <= 0):
        raise ValueError("dimension must be a positive integer")

    embedder_config = _model_dict(embedder)
    if not isinstance(embedder_config.get("provider"), str) or not embedder_config["provider"]:
        raise ValueError("embedder.provider is required")
    if distance_metric not in (None, "l2_squared", "inner_product", "cosine"):
        raise ValueError("distance_metric is invalid")

    enrichments: list[dict[str, Any]] = []
    for index, source in enumerate(sources):
        if not source.field and not source.template:
            raise ValueError(f"sources[{index}] requires field or template")
        if source.source_artifact == "":
            raise ValueError(f"sources[{index}].source_artifact cannot be empty")
        enrichment: dict[str, Any] = {
            "name": source.artifact,
            "kind": "embedding",
        }
        if source.field:
            enrichment["field"] = source.field
        if source.template is not None:
            enrichment["template"] = source.template
        if source.source_artifact is not None:
            enrichment["source_artifact_name"] = source.source_artifact
        if dimension is not None:
            enrichment["expected_dims"] = dimension
        if vector_space is not None:
            if not vector_space:
                raise ValueError("vector_space cannot be empty when provided")
            enrichment["vector_space"] = vector_space
        enrichments.append(enrichment)

    result: dict[str, Any] = {
        "name": name,
        "type": "embeddings",
        "sources": artifact_index_sources(*(source.artifact for source in sources)),
        "enrichments": enrichments,
        "embedder": embedder_config,
    }
    if sparse:
        result["sparse"] = True
    if dimension is not None:
        result["dimension"] = dimension
    if distance_metric is not None:
        result["distance_metric"] = distance_metric
    return result
